blog/_posts/2017-12-05-ml&pl.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: post
+title:  Programming Languages for Machine Learning
+---
+
+> Any sufficiently complicated machine learning system contains an ad-hoc, informally-specified, bug-ridden, slow implementation of half of a programming language.
+
+As programming languages (PL) people, we have watched with great interest as machine learning (ML) has exploded – and with it, the complexity of ML models and the frameworks people are using to build them. It’s clear that models are increasingly *programs* with support for programming constructs like loops and recursion, and this brings out many interesting issues in the tools we use to create them – that is, programming languages.
+
+While machine learning does not yet have a dedicated language, several efforts are creating hidden new languages underneath a Python API (like TensorFlow) while others are reusing Python as a modelling language (like PyTorch). We’d like to ask – are new ML-tailored languages required, and if so, why? More importantly, what might the ideal ML language of the future look like?
+
+## Pig Latin, and Other Hidden Languages
+
+TensorFlow (TF) and its ilk[^tf] are [already programming languages](https://dl.acm.org/citation.cfm?doid=3088525.3088527), albeit limited ones. This may seem surprising given that one uses Python to program TF. However, consider that TF requires you to write Python code to [build an expression tree](https://www.tensorflow.org/programmers_guide/graphs) in its internal language, which it then evaluates.
+
+[^tf]: We use TensorFlow for example, but could substitute other "define-before-run" frameworks like CNTK or MXNet.
+
+In fact, you can program in "lazy" TensorFlow style in any language. Consider the following JavaScript code, which implements a trivial function (`add`) in this style:
+
+{% highlight julia %}
+function add(a,b) {
+  return `${a}+${b}`;
+}
+x = 1; y = 2
+z = add(‘x’, ‘y’) // ‘x+y’
+eval(z) // 3
+
+x = 4
+eval(z) // 6
+{% endhighlight %}
+
+Here we are *metaprogramming* – writing code that writes code. In this case both the meta-language and the target language are the same (JavaScript) but they could just as well be different languages (as in the C preprocessor for the C language), or we can use a data structure (an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)) instead of a string – the principle is the same. In TensorFlow, Python serves as a meta-language for writing programs in TF’s graph-based language[^ast]. If you’re not convinced, consider that TensorFlow’s graph even supports constructs like variable scoping and control flow – but rather than using Python syntax, you manipulate these constructs through an API.
```

Comment:
Show some example code for the control flow. You're assuming the reader knows how ugly TF control constructs look.

Reply:
We haven't assumed (or argued) that TF is ugly here – the point isn't to be anti-TF but just to describe how it's like a language, so I think this suffices (although I've added some links for the interested reader).

Reply:
I think that you should put a code snippet of what it looks like so the reader can decide for themselves whether the feel is "good" or not. I think you hit home the fact that these pieces are required, but haven't given a good feel for what this actually means to code in these frameworks, which can be done by showing the custom control constructs that things like TF use. I'm not saying you should argue that TF is ugly here, just show its code and the reader will know (or make their own judgement, but I assume the reader is wise).